Fork Docs Update (lava-testnet-2) (#151)
lavanet/docs: docs/provider/provider-tls.md

<details> <summary> 🖧 Multiple Records (Recommended) </summary>

We recommend you create a separate `A-Record` for each one of the chains that you plan to support. This is more secure, as the default behavior is to refuse connection unless a consumer connects on the correct subdomain. 

For each chain you want to support, add an `A-Record` with the desired chain name as the `Host`, the `Value` will be your server IP.
For example, if you wanted to support Ethereum & Lava Mainnets, Your DNS Settings should include the following :

| Record Type       |  Host | Value |
| ------------------|-------| ----- |
| A-Record          |   eth |   Your-Server-Public-IP-Address-Here   |
| A-Record          |  lava |   Your-Server-Public-IP-Address-Here   |

</details>

<details> <summary> ⚀ Single Record </summary>

Alternatively, you can create one `A-Record` that captures traffic to all sub-domains. If you are supporting a large number of chains that frequently changes, doing this may somewhat simplify your process.

| Record Type       | Host | Value |
|-------------------|------|-------|
| A-Record          | *    | Your-Server-Public-IP-Address-Here |

</details>
<br />

> ### 📂 Update and Install Required Packages 
> 
> On the server which you would like to turn into a provider, run updates and install the following packages:
> 
> ```bash
> sudo apt-get update
> sudo apt install certbot net-tools nginx python3-certbot-nginx -y

Reviewer: A few comments: `apt update` does not update the system. This command updates the package database. `apt upgrade` is the command to update the system. "run updates" can be misleading in this case, hence my suggestion below. Also, `apt` is a convenient wrapper around `apt-get` for interactive usage.

```suggestion
### 📂 Install Required Dependencies

We will guide you on setting up and configuring Nginx to use a TLS certificate and handle connections to different provider processes. It is also possible to use alternative solutions for these two tasks, such as Caddy, Envoy or your preferred solution.

Run the following commands to install the required packages:

```bash
sudo apt update
sudo apt install certbot net-tools nginx python3-certbot-nginx -y
```

Author: Good catch. Adding your change!

```

> 
> This guide uses `nginx` to handle routing to different provider processes, but you can also use Caddy, Envoy or your preferred solution.

Reviewer: ```suggestion
```

This was stated above. Stating other options right in the beginning makes the guide less opinionated and therefore more friendly :)

<br />

### 📮 Generate Certificate 

> Next, we need to actually create the `TLS certificate` via the certifying authority. This process is automated by `certbot`.

Reviewer: Do you know what are the requirements for the certificate? Are there any specific ciphers and version that people should use? I can provide suggestions in case this is not set yet.

Author: Unfortunately, I am not aware of any specifics. The fact that certbot generated certs work is the extent of my knowledge at this time. Your suggestins are welcome!

Author: To revisit once @ranlavanet is back in office.

```

You'll need both `Certificate Path` and `Private Key Path` for your next step.

<br />

### 🗃️ Add an Nginx Config for Each Domain

Lava recommends running each chain under a separate provider process. This will separate error logs and protect against complete provider failure in the case of a problematic provider process. The first step of this is to create different nginx routes for each chain.

For each chain that you want to support, you will need to create a separate `nginx` config file. 
`cd` into `/etc/sites-available` and create a `server` file for each chain. You will need to select an open port for each chain. `Nginx` will use these config files to create your routes.

<Tabs>
<TabItem value="eth_ex" label="eth_server">

🟢 ```sudo nano eth_server``` 

```
server {
    listen 443 ssl http2;
    server_name eth.your-site.com;

    ssl_certificate /etc/letsencrypt/live/your-site.com/fullchain.pem;
    ssl_certificate_key /etc/letsencrypt/live/your-site.com/privkey.pem;
    error_log /var/log/nginx/debug.log debug;

    location / {
        proxy_pass http://127.0.0.1:2223;
        grpc_pass 127.0.0.1:2223;
    }

> }

Reviewer: Is there a reason to not have a redirect from port 80 to 443?

Author: I'm not certain- I think @ranlavanet had a reason but he's OOO. What do you suggest?

Reviewer: Well, it was stated that secure connections is mandatory, but it is customary to have an "upgrade" from insecure to secure connection (3xx redirect from :80 to :443). Not critical, but nice to have. We can move forward without it.

Author: Purely speculation, but may have something to do with how Nginx handles insecure requests. TBD ✅

```

</TabItem>
<TabItem value="lav_ex" label="lava_server">

🟢 ```sudo nano lava_server```

```
server {
    listen 443 ssl http2;
    server_name lava.your-site.com;

    ssl_certificate /etc/letsencrypt/live/your-site.com/fullchain.pem;
    ssl_certificate_key /etc/letsencrypt/live/your-site.com/privkey.pem;
    error_log /var/log/nginx/debug.log debug;

    location / {
        proxy_pass http://127.0.0.1:2224;
        grpc_pass 127.0.0.1:2224;
    }
}
```

</TabItem>
</Tabs>

:::tip

The above examples use ports `2223` and `2224`, respectively. You can choose any port that works for your purposes. 

:::

<br />

### ♻️ Restart Nginx

You will need to refresh the Nginx server:

```bash

> sudo systemctl nginx restart

Reviewer: The correct order is the action before the service:

```suggestion
sudo systemctl restart nginx
```

`systemctl nginx start` will fail and give this error: `Unknown command verb nginx.service.`

Author: Typo. Good catch!

```
<br />

### 🧪 Test Nginx Configuration

Now, ensure that your `nginx` setup is working! ✅

```bash
sudo nginx -t
```

🖳 Expected Output:
```
nginx: the configuration file /etc/nginx/nginx.conf syntax is ok
nginx: configuration file /etc/nginx/nginx.conf test is successful
```

<br />

The syntax on your files must be precise. Misplaced or invisible characters or inconsistent indentation can cause errors.

> 

Reviewer: What do you think about moving this section to before restarting Nginx? If there's an error/typo in the Nginx configuration, the previous section will error. Checking the configuration before restarting/reloading will allow people to fix mistakes before moving to the next section.

Also, Nginx does not care about indentation. That won't give an error. Maybe you meant "inconsistent curly braces"?

Author: I think it's a good idea to move it. I will change and link to the commit here!

My mistake on the indentation - I was thinking of the yaml files that are fed to start the rpcprovider process. I think that whole comment was meant to be applied to them. I will move to appropriate place.

Author: ✅ Fixed 
Commit [d858ca0e5c0384a43dca7c9f45a9fba7c7439a2d](https://github.com/lavanet/docs/pull/151/commits/d858ca0e5c0384a43dca7c9f45a9fba7c7439a2d)

        address: 127.0.0.1:2224
        disable-tls: true
      node-urls:
        - url: ws://127.0.0.1:26657/websocket
        - url: http://127.0.0.1:26657
    - api-interface: grpc
      chain-id: LAV1
      network-address:
        address: 127.0.0.1:2224
        disable-tls: true
      node-urls: 
        url: 127.0.0.1:9090
    - api-interface: rest
      chain-id: LAV1
      network-address:
        address: 127.0.0.1:2224
        disable-tls: true
      node-urls: 
        url: http://127.0.0.1:1317
```

</TabItem>
<TabItem value="eth_yml" label="eth-provider">

`nano eth-provider.yml`

```yaml

endpoints:
    - api-interface: jsonrpc
      chain-id: ETH1
      network-address:
        address: 127.0.0.1:2223
        disable-tls: true
      node-urls: 
        url: wss://ethereum-rpc.com/ws/
```

</TabItem>
</Tabs>

Once we've created these files we can move onto starting the processes!
<br />

### 🏁 Start the Provider Process(es)

First, we need to verify the blockchain is reachable with `lavad status` ✅ Once we have successfully gotten node info, we can continue. If you have not already set up your node you need to do so now. Here is where rubber meets road!

> In this example, we use the built-in terminal multiplexer `screen` to run multiple provider processes. Begin by typing `screen`. 

Reviewer: ```suggestion
In this example, we use the built-in terminal multiplexer `screen` to run multiple provider processes. Begin by typing `screen`. But you can also use a different multiplexer, e.g. `tmux`.
```

Author: Unopinionated :)
